src/ucp/wireup/ep_match.c

```diff
@@ -0,0 +1,205 @@
+/**
+ * Copyright (C) Mellanox Technologies Ltd. 2018.  ALL RIGHTS RESERVED.
+ *
+ * See file LICENSE for terms.
+ */
+
+#include <ucp/core/ucp_ep.h>
+#include <ucp/core/ucp_ep.inl>
+#include <ucp/wireup/ep_match.h>
+#include <inttypes.h>
+
+
+__KHASH_IMPL(ucp_ep_match, static UCS_F_MAYBE_UNUSED inline, uint64_t,
+             ucp_ep_match_entry_t, 1, kh_int64_hash_func, kh_int64_hash_equal);
+
+
+#define ucp_ep_match_list_for_each(_elem, _head, _member) \
+    for (_elem = ucs_container_of((_head)->next, typeof(*_elem), _member); \
+         &(_elem)->_member != NULL; \
+         _elem = ucs_container_of((_elem)->_member.next, typeof(*_elem), _member))
+
+static inline void ucp_ep_match_list_add_tail(ucs_list_link_t *head,
+                                              ucs_list_link_t *elem)
+{
+    ucs_list_link_t *last;
```

Comment:
why not just `ucs_list_add_tail`?

Reply:
because this is not a circular list (see https://github.com/openucx/ucx/pull/2500/files/23d63ccf81821b049bf1faa8a3afabc9d1751376#diff-7164313e3a0d83529b5a237714cc6da1R18)
the list cannot be circular because the hash can grow and the list head will move in memory, so the list head cannot be part of the list